quinn-proto/src/connection/mod.rs

```diff
@@ -514,35 +514,18 @@ where
                 Some(self.populate_packet(space_id, &mut buf, buf_capacity - builder.tag_len))
             };
 
-            let exact_number = builder.exact_number;
-            let (size, padded) = self.finish_packet(builder, &mut buf);
-
-            if let Some(mut sent) = sent {
-                sent.padding = padded;
+            if let Some(sent) = &sent_frames {
                 // If we sent any acks, don't immediately resend them. Setting this even if ack_only is
                 // false needlessly prevents us from ACKing the next packet if it's ACK-only, but saves
                 // the need for subtler logic to avoid double-transmitting acks all the time.
+                // This reset needs to happen before we check whether more data
+                // is available in this space - because otherwise it would return
+                // `true` purely due to the ACKs.
                 self.spaces[space_id].permit_ack_only &= sent.acks.is_empty();
-
-                self.on_packet_sent(
-                    now,
-                    space_id,
-                    exact_number,
-                    SentPacket {
-                        acks: sent.acks,
-                        time_sent: now,
-                        size: if sent.padding || ack_eliciting {
-                            size as u16
-                        } else {
-                            0
-                        },
-                        ack_eliciting,
-                        retransmits: sent.retransmits,
-                        stream_frames: sent.stream_frames,
-                    },
-                );
             }
 
+            self.finish_and_track_packet(now, builder, sent_frames.take(), &mut buf);
```

Comment:
Yes. In this commit. But not anymore after the factor to GSO, since the `Option` is kept around outside the send loop for the case a previous packet needs to be finalized.

Reply:
Will be fixed in the next intermediate commit